### ISSUE TITLE
Rowan Electronic Record T2 Week 6

### ISSUE THREAD

**Issue by r-preston** [r-preston/MPhysProject2021#47]
# Chi-Square test
Comparison script now preforms a Chi-Square goodness of fit test between the meaureed and predicted differential cross section in bins of `y`. [commit](https://github.com/r-preston/MPhysProject2021/commit/42b2c2fdc50f534d89c53d56547bb660148578b6)

Output:
```json
{
    "chi2": 31.125191627417845,
    "k": 9,
    "significance": 0.001,
    "critical_value": 27.877,
    "passed": 1
}
```
# Varying model parameters

Next looked at investigating how changing various parameters of the Pythia simulation affects the fit against the measured distributions.
Of particular interest is variation in the rapdity plot with variations of the PDF set used, and variation in the transverse momentum plot with variations in the strong coupling constant α_s.

### Pythia code
Modified `generate_z_events.cpp` to produce a tree of `Z->mu+mu-` for several variations of both PDF set and alpha_s values. The trees are written to a ROOT file in a seperate `TDirectory` for each parameter change. Data about the parameter variations is written to `data/prediction_parameters.json`.

### Analysis code
The analysis code has also been modified to accomodate several Pythia runs with different variables. The cross-section is and Chi-Square against measurement is still calculated for the 'default settings' Pythia run, however the code now also performs Chi-Square tests in both pT and rapidity for each changed Pythia variable, as well as plotting the cross sections for each simulation run against each other.

<img src="https://raw.githubusercontent.com/r-preston/MPhysProject2021/master/rowan2021-22/progress/18-02/variations.png" width="100%">

Example output for Chi-Square tests with different PDF sets (stored as CSV data):

value | Chi2 (pT) | Chi2 (y)
---|---|---
`NNPDF2.3` | 19.52142874430374 | 20.974095075018553
`CT09MCS` | 27.486521143973356 | 22.13301971537484
`NNPDF3.1` | 35.15066187287158 | 33.35963701406377

**Comment by mvesteri** [comment]
This chi2 looks plausible. 

As we discussed last week, an interesting next step would be to now produce Pythia predictions with a few other PDF sets and produce the corresponding data/model plots and chi2 values.

**Comment by r-preston** [comment]
@mvesteri I've modified my code to allow running easily running Pythia multiple times with different configuration. I've also changed the analysis code to compare the different Pythia runs and perform Chi-Square tests on them.

It ended up being rather more involved and long-winded than I'd anticipated but I'm quite pleased with the end result. It should be easier in future to add and tweak the different Pythia runs
